src/SMCE/BoardView.cpp

```diff
@@ -298,43 +298,232 @@ bool FrameBuffer::read_rgb888(std::span<std::byte> buf) {
+        *target++ = (left << 4) | (left & (std::byte)0b00001111);   // blue
+    }
+}
+
+// buf has the RGB444 format
+// framebuffer always has RGB888 format
+bool FrameBuffer::write_rgb444(std::span<const std::byte> source) {
     if (!exists())
         return false;
 
     auto& frame_buf = m_bdat->frame_buffers[m_idx];
-    if (buf.size() != frame_buf.data.size() / 2)
+    if (source.size() != frame_buf.data.size() / 3 * 2)
         return false;
 
     [[maybe_unused]] std::lock_guard lk{frame_buf.data_mut};
 
-    auto* to = frame_buf.data.data();
-    for (std::byte from : buf) {
-        *to++ = from & std::byte{0xF};
-        *to++ = from << 4; // Might be a bug there in the case where we have an odd number of pixels in the frame
+    // to is a pointer to the frame buffer data (a long array of bytes, where every 4 bytes is one pixel)
+    std::byte* target = frame_buf.data.data();
+    convert_rgb444_to_rgb888(source, target);
+    return true;
+}
+
+// convert 3 bytes into 2 bytes
+void convert_rgb888_to_rgb444(const std::byte* source, std::span<std::byte> target) {
+
+    for (auto i = target.begin(); i != target.end(); ++i) {
+        // read red/green/blue
+
+        // pick values from the 24bits from frame buffer
+        std::byte red = *source++;
+        std::byte green = *source++;
+        std::byte blue = *source++;
+
+        // write to the target
+        *i++ = (green & (std::byte)0b11110000) | ((blue & (std::byte)0b11110000) >> 4);
+        *i = (red >> 4);
     }
+}
+// read from the frame buffer (rgb888) into the buf (rgb444)
+bool FrameBuffer::read_rgb444(std::span<std::byte> target) {
+    if (!exists())
+        return false;
+
+    auto& frame_buf = m_bdat->frame_buffers[m_idx];
+    if (target.size() != frame_buf.data.size() / 3 * 2)
+        return false;
+    [[maybe_unused]] std::lock_guard lk{frame_buf.data_mut};
 
+    const auto* source = frame_buf.data.data();
+    convert_rgb888_to_rgb444(source, target);
     return true;
 }
 
-bool FrameBuffer::read_rgb444(std::span<std::byte> buf) {
+void convert_rgb565_to_rgb888(std::span<const std::byte> source, std::byte* target) {
+    // read two bytes at the time
+    for (auto i = source.begin(); i != source.end(); ++i) {
+        std::byte left = *i++;
+        std::byte right = *i;
+        *target++ = (right & (std::byte)0b11111000) | (right >> 5); // red
+        *target++ = ((right & (std::byte)0b00000111) << 5) | ((left & (std::byte)0b11100000) >> 3) |
+                    ((right & (std::byte)0b00000110) >> 1);           // green
+        *target++ = (left << 3) | ((left & (std::byte)0b11100) >> 2); // blue
+    }
+}
+// Implementation of rgb565. conversion from RGB888 to RGB565
+bool FrameBuffer::write_rgb565(std::span<const std::byte> source) {
     if (!exists())
         return false;
 
     auto& frame_buf = m_bdat->frame_buffers[m_idx];
-    if (buf.size() != frame_buf.data.size())
+    if (source.size() != frame_buf.data.size() / 2)
```

Comment:
Why this is size()/2. I assume that the `frame_buf.data.size()==24`, is that right?

Reply:
aha rgb888 is 32bits, 8 empty bits in the begining